vm/src/obj/objiter.rs

```diff
@@ -152,25 +149,44 @@ impl PyValue for PySequenceIterator {
 
 #[pyimpl]
 impl PySequenceIterator {
+    pub fn new_forward(obj: PyObjectRef) -> Self {
+        Self {
+            position: AtomicCell::new(0),
+            obj,
+            reversed: false,
+        }
+    }
+
+    pub fn new_reversed(obj: PyObjectRef, len: isize) -> Self {
+        Self {
+            position: AtomicCell::new(len - 1),
+            obj,
+            reversed: true,
+        }
+    }
+
     #[pymethod(name = "__next__")]
     fn next(&self, vm: &VirtualMachine) -> PyResult {
-        if self.position.get() >= 0 {
-            let step: isize = if self.reversed { -1 } else { 1 };
-            let number = vm.ctx.new_int(self.position.get());
-            match vm.call_method(&self.obj, "__getitem__", vec![number]) {
-                Ok(val) => {
-                    self.position.set(self.position.get() + step);
-                    Ok(val)
-                }
-                Err(ref e) if objtype::isinstance(&e, &vm.ctx.exceptions.index_error) => {
-                    Err(new_stop_iteration(vm))
+        let mut prev = self.position.load();
```

Comment:
I might be missing something but I think we can go for a simpler solution here:
```rust
let step = if self.reversed { -1 } else { 1 };
let curr_pos = self.position.fetch_add(step);
if curr_pos >= 0 {
...
```
then use the same code as before.

This will answer what we expect from an iterator. Each time next is called we will receive a unique item from the iterable. Once the iterator is exhausted all consecutive calls will get stop iteration.

Reply:
Ah, that probably would work since it's an isize and it wouldn't matter if it underflowed.